base/acme/src/org/dogtagpki/acme/server/ACMEEngine.java

```diff
@@ -222,4 +239,128 @@ public void contextDestroyed(ServletContextEvent event) {
         }
     }
+
+    public ACMENonce createNonce() throws Exception {
+
+        ACMENonce nonce = new ACMENonce();
+
+        // generate 128-bit nonce with JSS
+        // TODO: make it configurable
+
+        byte[] bytes = new byte[16];
+        SecureRandom random = SecureRandom.getInstance("pkcs11prng", "Mozilla-JSS");
+        random.nextBytes(bytes);
+        String value = Base64.encodeBase64URLSafeString(bytes);
+
+        nonce.setValue(value);
+
+        // set nonce to expire in 30 minutes
+        // TODO: make it configurable
+
+        long currentTime = System.currentTimeMillis();
+        long expirationTime = currentTime + 30 * 60 * 1000;
+
+        nonce.setExpirationTime(new Date(expirationTime));
+
+        database.addNonce(nonce);
+        logger.info("Nonce: " + nonce);
+
+        return nonce;
+    }
+
+    public void validateNonce(String value) throws Exception {
+
+        ACMENonce nonce = database.removeNonce(value);
+
+        if (nonce == null) {
+            throw new Exception("Invalid nonce: " + value);
+        }
+
+        long currentTime = System.currentTimeMillis();
+        long expirationTime = nonce.getExpirationTime().getTime();
+
+        if (expirationTime <= currentTime) {
+            throw new Exception("Expired nonce: " + value);
+        }
+
+        logger.info("Valid nonce: " + value);
+    }
+
+    public void purgeNonces() throws Exception {
+        database.removeExpiredNonces(new Date());
+    }
+
+    public void validateJWS(JWS jws, String alg, JWK jwk) throws Exception {
+
+        logger.info("Validating " + alg + " JWS");
+
+        String message = jws.getProtectedHeader() + "." + jws.getPayload();
+        byte[] signature = Base64.decodeBase64(jws.getSignature());
+
+        // JWS validation based on RFC 7515
+        // TODO: support other algorithms
+
+        if ("RS256".equals(alg)) {
+
+            String kty = jwk.getKty();
+            KeyFactory keyFactory = KeyFactory.getInstance(kty, "Mozilla-JSS");
+
+            String n = jwk.getN();
+            BigInteger modulus = new BigInteger(1, Base64.decodeBase64(n));
+
+            String e = jwk.getE();
+            BigInteger publicExponent = new BigInteger(1, Base64.decodeBase64(e));
+
+            RSAPublicKeySpec keySpec = new RSAPublicKeySpec(modulus, publicExponent);
+            PublicKey publicKey = keyFactory.generatePublic(keySpec);
+
+            Signature signer = Signature.getInstance("SHA256withRSA", "Mozilla-JSS");
+            signer.initVerify(publicKey);
+            signer.update(message.getBytes());
+
+            if (!signer.verify(signature)) {
+                throw new Exception("Invalid " + alg + " JWS");
+            }
+
+        } else {
+            throw new Exception("Unsupported JWS algorithm: " + alg);
+        }
+    }
+
+    public String generateHash(String data) throws Exception {
+        MessageDigest digest = MessageDigest.getInstance("SHA-256", "Mozilla-JSS");
```

Comment:
Does this need to be configurable in the future?

Reply:
Possibly. The current `generateHash()` will be used for both JWS thumbprint and dns-01 validation. I think they should be separate so they can be configured independently. Let me revise the patch.

Reply:
(To be clear, I'm fine with just marking this with `// TODO` like in the earlier section of code).

Reply:
Yes, it will be a TODO for now, but there will be separate code for JWS thumbprint and dns-01 validation :)